src/style_layer.js

```diff
@@ -0,0 +1,100 @@
+
+    options: {
+        sanitizer: require('sanitize-caja')
+    },
+
+    initialize: function(_, options) {
+        L.TileLayer.prototype.initialize.call(this, undefined, options);
+
+        this.options.tiles = this._formatTileURL(_);
+
+        this._tilejson = {};
+
+        this._getAttribution(_);
+    },
+
+    _setTileJSON: function(json) {
+        util.strict(json, 'object');
+
+        L.extend(this.options, {
+            attribution: this.options.sanitizer(json.attribution),
+            tms: false,
+            tileSize: 512,
+            zoomOffset: -1
+        });
+
+        this._tilejson = json;
+        this.redraw();
+        return this;
+    },
+
+    _getAttribution: function(_) {
+        var styleURL = format_url.style(_, this.options && this.options.accessToken);
+        request(styleURL, L.bind(function(err, style) {
+            if (err) {
+                util.log('could not load Mapbox style at ' + styleURL);
+                this.fire('error', {error: err});
+            }
+            var sources = [];
+            for (var id in style.sources) {
+                var source = style.sources[id].url.split('mapbox://')[1];
+                sources.push(source);
+            }
+            request(format_url.tileJSON(sources.join(), this.options.accessToken), L.bind(function(err, json) {
+                if (err) {
+                    util.log('could not load TileJSON at ' + _);
+                    this.fire('error', {error: err});
+                } else if (json) {
+                    this._setTileJSON(json);
+                    this.fire('ready');
+                }
+            }, this));
+        }, this));
+    },
+
+    // disable the setUrl function, which is not available on mapbox tilelayers
+    setUrl: null,
+
+    _formatTileURL: function(style) {
+        var retina = L.Browser.retina ? '@2x' : '';
+        if (typeof style === 'string') {
+            if (style.indexOf('mapbox://styles/') === -1) {
+                util.log('Incorrectly formatted Mapbox style at ' + style);
+                this.fire('error');
+            }
+            var ownerIDStyle = style.split('mapbox://styles/')[1];
+            return format_url('/styles/v1/' + ownerIDStyle + '/tiles/{z}/{x}/{y}' + retina, this.options.accessToken);
+        } else if (typeof style === 'object') {
+            return format_url('/styles/v1/' + style.owner + '/' + style.id + '/tiles/{z}/{x}/{y}' + retina, this.options.accessToken);
+        }
+    },
+
+    // this is an exception to mapbox.js naming rules because it's called
+    // by `L.map`
+    getTileUrl: function(tilePoint) {
+        var templated = L.Util.template(this.options.tiles, tilePoint);
+        return templated;
+    },
+
+    // TileJSON.TileLayers are added to the map immediately, so that they get
+    // the desired z-index, but do not update until the TileJSON has been loaded.
+    _update: function() {
+        if (this.options.tiles) {
```

Comment:
Since the constructor sets `this.options.tiles`, this condition will always be true.

Instead, let's remove the asynchronicity around tile properties, and figure out how to refresh just the attribution when we get it:

* Move setting `tileSize: 512` and `zoomOffset: -1` to the constructor.
* Inline the rest of `_setTileJSON` into `_getAttribution`.
* Figure out what needs to happen inside `_getAttribution` to refresh the attribution when we get it.